storage/s3/src/main/java/io/aiven/kafka/tieredstorage/storage/s3/S3Storage.java

```diff
@@ -46,9 +46,23 @@ public void configure(final Map<String, ?> configs) {
     }
 
     @Override
-    public void upload(final InputStream inputStream, final String key) throws StorageBackendException {
-        try (final var out = s3OutputStream(key)) {
-            inputStream.transferTo(out);
+    public long upload(final InputStream in, final String key) throws StorageBackendException {
+        try {
+            final var out = s3OutputStream(key);
+            transferContent(key, in, out);
+            return out.uploadedBytes();
+        } catch (final AmazonS3Exception e) {
+            throw new StorageBackendException("Failed to initiate uploading " + key, e);
+        }
+    }
+
+    void transferContent(final String key,
+                         final InputStream in,
+                         final S3MultiPartOutputStream out)
+        throws StorageBackendException {
+        try {
+            in.transferTo(out);
+            out.close();
```

Comment:
We can add one line to `upload` to make it return the size:
```java
public long upload(final InputStream inputStream, final String key) throws StorageBackendException {
    try (final var out = s3OutputStream(key)) {
        inputStream.transferTo(out);
        return out.uploadedBytes();
    } catch (final AmazonS3Exception | IOException e) {
        throw new StorageBackendException("Failed to upload " + key, e);
    }
}
```

Anything I miss that make it necessary to split it like this and abandon try-with-resources?

Reply:
In the meantime: I was counting bytes on flush which depended on close (last flush happens there). But I have moved the increment a bit earlier to the write method, so we can get the size as it's transferred to the buffer but not flushed necessarily.